### PR TITLE
[RISCV] Add initial support of memcmp expansion

### PR DESCRIPTION
**Pull request by wangpc-pp**
There are two passes that have dependency on the implementation
of `TargetTransformInfo::enableMemCmpExpansion` : `MergeICmps` and
`ExpandMemCmp`.

This PR adds the initial implementation of `enableMemCmpExpansion`
so that we can have some basic benefits from these two passes.

We don't enable expansion when there is no unaligned access support
currently because there are some issues about unaligned loads and
stores in `ExpandMemcmp` pass. We should fix these issues and enable
the expansion later.

Vector case hasn't been tested as we don't generate inlined vector
instructions for memcmp currently.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:2122]: Is one supposed to be enableUnalignedVectorMem()?

**Comment by lukel97** [comment]
Looks like some SPEC benchmarks use memcmp, I'm benchmarking them with this patch now. I'll update you when it finishes :)

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:2128]: I checked what this is used for, and it seems like we'll try to emit a 3 byte ld/st?  That seems wrong.

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (preames) [inline llvm/test/CodeGen/RISCV/memcmp.ll:9]: Please add coverage for the following cases:
* Length = 0, 1, 2, 3, 4, 5, 6, 7,8,16, 32, 64, runtime
* memcmp *without* comparison against zero
* RUN lines for vector

Once done, please land the tests so that we can see current expansions.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:2122]: Unless I'm missing something, there is no check for sufficient alignment on *any* load generated by ExpandMemCmp, as a result, we can only legally provide a non-byte size if unaligned access is fully supported.

(This is really bug in ExpandMemCmp - it should be asking if the load is properly aligned.)

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: https://github.com/llvm/llvm-project/pull/107824

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: Yes, it seems we will generate i40/i48/... loads (see also the tests in https://github.com/llvm/llvm-project/pull/70469).
This may not work for RISC-V.

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp]: I removed it first. If RISC-V can benefit from it, we can add it back as a follow-up.

**Comment by asb** [comment]
This is perhaps more of a comment for #107824 than for this one, but I think we'd benefit from some level of test coverage for OptForSize to demonstrate that we stick with the libcall in cases where expanding increases code size.

**Comment by wangpc-pp** [comment]
> This is perhaps more of a comment for #107824 than for this one, but I think we'd benefit from some level of test coverage for OptForSize to demonstrate that we stick with the libcall in cases where expanding increases code size.

Thanks! Done!

**Comment by asb** [comment]
> > This is perhaps more of a comment for #107824 than for this one, but I think we'd benefit from some level of test coverage for OptForSize to demonstrate that we stick with the libcall in cases where expanding increases code size.
> 
> Thanks! Done!

Cheers! It strikes me that we should probably be less aggressive about expanding memcmp for at least `Os` without unaligned access support.  I'm not quite sure where the cutoff should be - `Os` of course has some leeway for increasing code size if we think it will benefit performance and I don't think we need to spend too much time trying to make a "perfect" decision when there's no single right answer. But e.g. bcmp_size_8 for Os without misaligned access support seems like more codesize blowup than I'd expect at that optimisation level. I'd welcome views on what a sensible threshold would be.

**Comment by lukel97** [comment]
I collected the stats on the number of memcmps that were inlined, it looks like we're able to expand a good chunk of them:
```
Program                                       expand-memcmp.NumMemCmpCalls              expand-memcmp.NumMemCmpInlined             
                                              lhs                          rhs    diff  lhs                            rhs    diff 
FP2017rate/510.parest_r/510.parest_r          410.00                       468.00 14.1%                                104.00  inf%
INT2017speed/602.gcc_s/602.gcc_s               83.00                        92.00 10.8%                                 36.00  inf%
INT2017rate/502.gcc_r/502.gcc_r                83.00                        92.00 10.8%                                 36.00  inf%
INT2017spe...00.perlbench_s/600.perlbench_s   207.00                       220.00  6.3%                                120.00  inf%
INT2017rat...00.perlbench_r/500.perlbench_r   207.00                       220.00  6.3%                                120.00  inf%
INT2017spe...ed/620.omnetpp_s/620.omnetpp_s   304.00                       306.00  0.7%                                 13.00  inf%
INT2017rate/520.omnetpp_r/520.omnetpp_r       304.00                       306.00  0.7%                                 13.00  inf%
FP2017rate/508.namd_r/508.namd_r               13.00                        13.00  0.0%                                 13.00  inf%
INT2017rate/541.leela_r/541.leela_r            40.00                        40.00  0.0%                                  3.00  inf%
INT2017speed/641.leela_s/641.leela_s           40.00                        40.00  0.0%                                  3.00  inf%
INT2017speed/625.x264_s/625.x264_s              8.00                         8.00  0.0%                                  6.00  inf%
INT2017spe...23.xalancbmk_s/623.xalancbmk_s     8.00                         8.00  0.0%                                  6.00  inf%
INT2017rate/557.xz_r/557.xz_r                   6.00                         6.00  0.0%                                  4.00  inf%
INT2017rat...23.xalancbmk_r/523.xalancbmk_r     8.00                         8.00  0.0%                                  6.00  inf%
INT2017rate/525.x264_r/525.x264_r               8.00                         8.00  0.0%                                  6.00  inf%
FP2017speed/644.nab_s/644.nab_s                77.00                        77.00  0.0%                                 71.00  inf%
FP2017speed/638.imagick_s/638.imagick_s         3.00                         3.00  0.0%                                            
FP2017rate/544.nab_r/544.nab_r                 77.00                        77.00  0.0%                                 71.00  inf%
FP2017rate/538.imagick_r/538.imagick_r          3.00                         3.00  0.0%                                            
FP2017rate/526.blender_r/526.blender_r         41.00                        41.00  0.0%                                 27.00  inf%
FP2017rate/511.povray_r/511.povray_r            5.00                         5.00  0.0%                                  5.00  inf%
INT2017speed/657.xz_s/657.xz_s                  6.00                         6.00  0.0%                                  4.00  inf%
```

There's a small difference in the number of original memcmp calls, there's some merge commits in this branch which might have changed the codegen slightly in the meantime.

I'm working on getting some runtime numbers now, sorry for the delay

**Comment by wangpc-pp** [comment]
> I'm working on getting some runtime numbers now, sorry for the delay.

No need to say sorry, I really appreciate your help! 😃

**Comment by lukel97** [comment]
The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r, no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!

**Comment by wangpc-pp** [comment]
> The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r on the BPI F3 (-O3 -mcpu=spacemit-x60), no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!

Thanks a lot! The result is within my expectation.
Is it OK to merge? The next step will be tuning for vectors.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:2125]: I wonder if this might be better

```
if (ST->is64Bit())
  Options.LoadSize = {8, 4, 2, 1};
else
  Options.LoadSize = {4, 2, 1};
```

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/memcmp-optsize.ll:107]: Would it be cheaper to Xor all the bytes individually and then Or the xor results together?

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/memcmp.ll:2931]: What is this code doing? It seems way more complicated than a 4 byte memcmp should be.

Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp.ll:2931]: I guess this is an expanded byteswap?

**Comment by topperc** [comment]
> The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r on the BPI F3 (-O3 -mcpu=spacemit-x60), no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!

Does spacemit-x60 support unaligned scalar memory and was your test with or without that enabled?

**Comment by wangpc-pp** [comment]
> > The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r on the BPI F3 (-O3 -mcpu=spacemit-x60), no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!
> 
> Does spacemit-x60 support unaligned scalar memory and was your test with or without that enabled?

It supports unaligned scalar but not unaligned vector. And it seems we don't add these features to `-mcpu=spacemit-x60`. So I think @lukel97 ran the SPEC without unaligned scalar.

**Comment by lukel97** [comment]
> > > The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r on the BPI F3 (-O3 -mcpu=spacemit-x60), no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!
> 
> > 
> 
> > Does spacemit-x60 support unaligned scalar memory and was your test with or without that enabled?
> 
> 
> 
> It supports unaligned scalar but not unaligned vector. And it seems we don't add these features to `-mcpu=spacemit-x60`. So I think @lukel97 ran the SPEC without unaligned scalar.

Yeah, -mno-strict-align gave a bus error. I ultimately built it without unaligned scalar since I wasn't sure if unaligned scalar was performant or not.

**Comment by wangpc-pp** [comment]
> > > > The run just finished, I'm seeing a 0.75% improvement on 500.perlbench_r on the BPI F3 (-O3 -mcpu=spacemit-x60), no regressions or improvements on the other benchmarks as far as I can see. Seems to check out with the number of memcmps inlined reported for perlbench!
> > 
> > > 
> > 
> > > Does spacemit-x60 support unaligned scalar memory and was your test with or without that enabled?
> > 
> > 
> > 
> > It supports unaligned scalar but not unaligned vector. And it seems we don't add these features to `-mcpu=spacemit-x60`. So I think @lukel97 ran the SPEC without unaligned scalar.
> 
> Yeah, -mno-strict-align gave a bus error. I ultimately built it without unaligned scalar since I wasn't sure if unaligned scalar was performant or not. 

IIRC, we have separated this into two options(scalar/vector) now. So maybe we can specify the scalar one.

**Comment by wangpc-pp** [comment]
Ping.

And some updates on vector support: currently, `ExpandMemCmp` will only generate integer types (`i128`, `i256`) and so on. So, if we simply add `128`, `256`, `vlen` to `LoadSizes`, the LLVM IR will use i128/i256/... and then they are expanded to legal scalar types as we don't mark i128/i256/... legal when RVV exists.

There are two ways to fix this:
1. Make `ExpandMemCmp` geenrate vector types/operations.
2. Make i128/i256/... legal.

I think the first one is the right approach but I need some agreements on this.

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp.ll:2931]: Add a RUN line with Zbb/Zbkb?

Maybe we should restrict MemCmp expansion to only when Zbb/Zbkb are are enabled?

**Comment by topperc** [comment]
> Ping.
> 
> And some updates on vector support: currently, `ExpandMemCmp` will only generate integer types (`i128`, `i256` and so on). So, if we simply add `128`, `256`, `vlen` to `LoadSizes`, the LLVM IR will use `i128`/`i256`/... and then they are expanded to legal scalar types as we don't mark `i128`/`i256`/... as legal when RVV exists.
> 
> There are two ways to fix this:
> 
> 1. Make `ExpandMemCmp` generate vector types/operations.
> 2. Make `i128`/`i256`/... legal on RISC-V.
> 
> I think the first one is the right approach but I need some agreements on this.

I think X86 only supports `IsZeroCmp` and has a pre-type legalization DAG combine to recognize the wide integer type. See X86ISelLowering `combineVectorSizedSetCCEquality`.

General memcmp is complicated with vector. You need to find the first element where the mismatch occurred and then compare only that element to find whether it is larger or smaller. I don't know if you can write that in target independent IR without it being really nasty and it wouldn't generate efficient code.

**Comment by wangpc-pp** [comment]
> > Ping.
> > And some updates on vector support: currently, `ExpandMemCmp` will only generate integer types (`i128`, `i256` and so on). So, if we simply add `128`, `256`, `vlen` to `LoadSizes`, the LLVM IR will use `i128`/`i256`/... and then they are expanded to legal scalar types as we don't mark `i128`/`i256`/... as legal when RVV exists.
> > There are two ways to fix this:
> > 
> > 1. Make `ExpandMemCmp` generate vector types/operations.
> > 2. Make `i128`/`i256`/... legal on RISC-V.
> > 
> > I think the first one is the right approach but I need some agreements on this.
> 
> I think X86 only supports `IsZeroCmp` for vector and has a pre-type legalization DAG combine to recognize the wide integer type. See X86ISelLowering `combineVectorSizedSetCCEquality`.

Thanks! I didn't notice that, I will try to implement this similarly in RISC-V.

> 
> General memcmp is complicated with vector. You need to find the first element where the mismatch occurred and then compare only that element to find whether it is larger or smaller. I don't know if you can write that in target independent IR without it being really nasty and it wouldn't generate efficient code.

Yes, I have noticed it. I was thinking that if we generate vector code earlier, we may get more spaces for optimizations. I will still try this approach, but I won't hold out high hopes.

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:107]: It costs the same number of instructions?

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: Done.
It seems we can benefit from not only `rev8` instructions but also `pack` instructions (for forming large integers).
But I don't think we should disable the expansion when Zbb/Zbkb don't exist. If these extensions are not supported and we don't expand memcmp, we would still execute these instructions in glibc, right?

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:187]: I was proposing

```
lbu a2, 1(a0)
lbu a0, 0(a0)
lbu a3, 1(a1)
lbu a1, 0(a1)
xor a2, a2, a3
xor a0, a0, a1
or  a2, a2, a0
snez a2, a2
```

that looks like fewer instructions.

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: The library call will do a byte loop without Zbb I think? I don't think it will do the byte swap with shift/and/or.

So the question is, is doing a wide load and an emulated byteswap cheaper than the byte loop?

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: Good question, I don't know the answer, but I think we can be aggressive here and enable it for all configurations. If we have some evidences that shows inefficiencies, we can go back to here and disable it. WDYT?

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:267]: Great! This is an optimization we can do in ExpandMemcmp, I will take a look.

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:267]: This seems to be more complicated than what I was thinking. For this case, we generate i16 loads/compares, and then we expand them since unaligned loads are illegal. We may expand memcmp to byte loop directly when there is no unaligned access.

**Comment by wangpc-pp** [comment]
Ping, any comment for current scalar part? I'm working on vector expansion and will post it in a few days.

File: llvm/test/CodeGen/RISCV/memcmp.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: Can we test the non-Zbb config on qemu and get the instruction count?

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp.ll]: Here is the code of memcmp copied from glibc: https://godbolt.org/z/4KxPTE6q1
There are many cases (which means branches) in this general implementation, at least we can benefit from unrolling and removal of branches.

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/memcmp-optsize.ll:4513]: What happened here?

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/memcmp-optsize.ll:5440]: Why didn't this use a lbu to avoid the andi later?

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:4513]: It expands to:
```
entry:
  %0 = load i32, ptr %s1, align 1
  %1 = load i32, ptr %s2, align 1
  %2 = icmp ne i32 %0, %1
  %3 = zext i1 %2 to i32
  %ret = icmp slt i32 %3, 0
  ret i1 %ret
```
And is optimized to :
```
entry:
  %0 = load i32, ptr %s1, align 1
  %1 = load i32, ptr %s2, align 1
  %2 = icmp ne i32 %0, %1
  %3 = zext i1 %2 to i32
  ret i1 false
```
This is how `bcmp` is expanded.

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:5440]: It seems to be related to https://reviews.llvm.org/D130397. Maybe we should revert this change or add a pattern to catch this.

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:4513]: I guess llvm is assuming that the bcmp should only be checked for eq/ne 0 and not gt/lt 0?

**Comment by topperc** [comment]
> At a macro level, it looks like ExpandMemCmp is making some problematic choices around unaligned loads and stores. As I commented before, ExpandMemCmp appears to be blindly emitting unaligned accesses (counted as one against budget) without accounting for the fact that such loads are going to be scalarized again (i.e. resulting in N x loads, where N is the type size). I think we need to fix this. In particular, the discussion around Zbb and Zbkb in this review seem to mostly come from cases where unaligned load.store are being expanded implicitly,
> 
> I don't believe this change should move forward until the underlying issue in ExpandMemCmp has been addressed.

Can we break this down into more manageable pieces? I think `enableUnalignedScalarMem() && (Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb() || IsZeroCmp)` might be a good starting point.

**Comment by preames** [comment]
> Can we break the enabling down into more manageable pieces? I think `enableUnalignedScalarMem() && (Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb() || IsZeroCmp)` might be a good starting point.

I'd be fine with this type of approach.

File: llvm/test/CodeGen/RISCV/memcmp-optsize.ll
Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/memcmp-optsize.ll:4513]: I don't know if this is the assumption of `bcmp` or just the behavior of `ExpandMemcmp`. I will keep these tests.

**Comment by wangpc-pp** [comment]
> > Can we break the enabling down into more manageable pieces? I think `enableUnalignedScalarMem() && (Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb() || IsZeroCmp)` might be a good starting point.
> 
> I'd be fine with this type of approach.

Thanks! I have applied this suggestion. For `ExpandMemcmp` issue, I will try to fix it later.

**Comment by topperc** [comment]
> > > Can we break the enabling down into more manageable pieces? I think `enableUnalignedScalarMem() && (Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb() || IsZeroCmp)` might be a good starting point.
> > 
> > 
> > I'd be fine with this type of approach.
> 
> Thanks! I have applied this suggestion. For `ExpandMemcmp` issue, I will try to fix it later.

I meant we should start with only doing inline memcmp expansion under those conditions. Not just the overlapping loads part.

**Comment by wangpc-pp** [comment]
> > > > Can we break the enabling down into more manageable pieces? I think `enableUnalignedScalarMem() && (Subtarget->hasStdExtZbb() || Subtarget->hasStdExtZbkb() || IsZeroCmp)` might be a good starting point.
> > > 
> > > 
> > > I'd be fine with this type of approach.
> > 
> > 
> > Thanks! I have applied this suggestion. For `ExpandMemcmp` issue, I will try to fix it later.
> 
> I meant we should start with only doing inline memcmp expansion under those conditions. Not just the overlapping loads part. The miscounting of loads that get split needs to be fixed before we can enable expansion without enableUnalignedScalarMem().

Oops, sorry I misunderstood it. Done.

**Review by topperc — APPROVED** [review]
LGTM, but wait for another review.

**Review by preames — APPROVED** [review]
LGTM

**Pull request merged**
